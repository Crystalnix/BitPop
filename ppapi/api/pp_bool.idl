/* Copyright (c) 2011 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/* This file defines the PP_Bool enumeration for use in PPAPI C headers. */

/* The PP_Bool enum is a boolean value for use in PPAPI C headers.  The
 * standard bool type is not available to pre-C99 compilers, and is not
 * guaranteed to be compatible between C and C++, whereas the PPAPI C
 * headers can be included from C or C++ code.
 */
enum PP_Bool {
  PP_FALSE = 0,
  PP_TRUE  = 1
};
