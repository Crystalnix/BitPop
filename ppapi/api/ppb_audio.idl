/* Copyright (c) 2011 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/* This file defines the audio API.
 */

/* Callback function type for SetCallback. */
typedef void PPB_Audio_Callback([out] mem_t sample_buffer,
                                [in] uint32_t buffer_size_in_bytes,
                                [inout] mem_t user_data);

/* Callback-based audio interface. User of audio must set the callback that will
 * be called each time that the buffer needs to be filled.
 *
 * A C++ example:
 *
 * void audio_callback(void* sample_buffer,
 *                     size_t buffer_size_in_bytes,
 *                     void* user_data) {
 *   ... fill in the buffer with samples ...
 * }
 *
 * uint32_t obtained;
 * AudioConfig config(PP_AUDIOSAMPLERATE_44100, 4096, &obtained);
 * Audio audio(config, audio_callback, NULL);
 * audio.StartPlayback();
 */
interface PPB_Audio_0_5 {
  /* Creates a paused audio interface. No sound will be heard until
   * StartPlayback() is called. The callback is called with the buffer address
   * and given user data whenever the buffer needs to be filled. From within the
   * callback, you should not call PPB_Audio functions.  The callback will be
   * called on a different thread than the one which created the interface. For
   * performance-critical applications (i.e. low-latency audio), the callback
   * should avoid blocking or calling functions that can obtain locks, such as
   * malloc. The layout and the size of the buffer passed to the audio callback
   * will be determined by the device configuration and is specified in the
   * AudioConfig documentation. If the configuration cannot be honored, or the
   * callback is null, the function returns 0.
   */
  PP_Resource Create(
      [in] PP_Instance instance,
      [in] PP_Resource config,
      [in] PPB_Audio_Callback audio_callback,
      [inout] mem_t user_data);

  /* Returns PP_TRUE if the given resource is an Audio resource, PP_FALSE
   * otherwise.
   */
  PP_Bool IsAudio(
      [in] PP_Resource resource);

  /* Get the current configuration. */
  PP_Resource GetCurrentConfig(
      [in] PP_Resource audio);

  /* Start the playback. Begin periodically calling the callback. If called
   * while playback is already in progress, will return PP_TRUE and be a no-op.
   * On error, return PP_FALSE.
   */
  PP_Bool StartPlayback(
      [in] PP_Resource audio);

  /* Stop the playback. If playback is already stopped, this is a no-op and
   * returns PP_TRUE. On error, returns PP_FALSE. If a callback is in progress,
   * StopPlayback will block until callback completes.
   */
  PP_Bool StopPlayback(
      [in] PP_Resource audio);
};
