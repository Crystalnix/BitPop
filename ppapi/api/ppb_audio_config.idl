/* Copyright (c) 2011 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/* This file defines the AudioConfig interface. */


/**
 * This enumeration contains audio frame count constants.
 * PP_AUDIOMINSAMPLEFRAMECOUNT is the minimum possible frame count.
 * PP_AUDIOMAXSAMPLEFRAMECOUNT is the maximum possible frame count.
 */
enum PP_AudioFrameSize {
  PP_AUDIOMINSAMPLEFRAMECOUNT = 64,
  PP_AUDIOMAXSAMPLEFRAMECOUNT = 32768
};


/**
 * PP_AudioSampleRate is an enumeration of the different audio sampling rates.
 * PP_AUDIOSAMPLERATE_44100 is the sample rate used on CDs and
 * PP_AUDIOSAMPLERATE_48000 is the sample rate used on DVDs and Digital Audio
 * Tapes.
 */
enum PP_AudioSampleRate {
  PP_AUDIOSAMPLERATE_NONE = 0,
  PP_AUDIOSAMPLERATE_44100 = 44100,
  PP_AUDIOSAMPLERATE_48000 = 48000
} ;


/* Interface for configuring audio output */
interface PPB_AudioConfig_0_5 {
  /* Create a 16 bit stereo config with the given sample rate. We guarantee
   * that PP_AUDIOSAMPLERATE_44100 and PP_AUDIOSAMPLERATE_48000 sample rates
   * are supported. The |sample_frame_count| should be the result of calling
   * RecommendSampleFrameCount. If the sample frame count or bit rate aren't
   * supported, this function will fail and return a null resource.
   *
   * A single sample frame on a stereo device means one value for the left
   * channel and one value for the right channel.
   *
   * Buffer layout for a stereo int16 configuration:
   * int16_t* buffer16;
   * buffer16[0] is the first left channel sample
   * buffer16[1] is the first right channel sample
   * buffer16[2] is the second left channel sample
   * buffer16[3] is the second right channel sample
   * ...
   * buffer16[2 * (sample_frame_count - 1)] is the last left channel sample
   * buffer16[2 * (sample_frame_count - 1) + 1] is the last right channel
   * sample
   * Data will always be in the native endian format of the platform.
   */
  PP_Resource CreateStereo16Bit(
      [in] PP_Instance instance,
      [in] PP_AudioSampleRate sample_rate,
      [in] uint32_t sample_frame_count);

  /* Returns a supported sample frame count closest to the given requested
   * count. The sample frame count determines the overall latency of audio.
   * Since one "frame" is always buffered in advance, smaller frame counts
   * will yield lower latency, but higher CPU utilization.
   *
   * Supported sample frame counts will vary by hardware and system (consider
   * that the local system might be anywhere from a cell phone or a high-end
   * audio workstation). Sample counts less than PP_AUDIOMINSAMPLEFRAMECOUNT
   * and greater than PP_AUDIOMAXSAMPLEFRAMECOUNT are never supported on any
   * system, but values in between aren't necessarily valid. This function
   * will return a supported count closest to the requested value.
   *
   * If you pass PP_AUDIOSAMPLERATE_NONE as the requested sample count, the
   * recommended sample for the local system is returned.
   */
  uint32_t RecommendSampleFrameCount(
      [in] PP_AudioSampleRate sample_rate,
      [in] uint32_t requested_sample_frame_count);

  /* Returns true if the given resource is an AudioConfig object. */
  PP_Bool IsAudioConfig(
      [in] PP_Resource resource);

  /* Returns the sample rate for the given AudioConfig resource. If the
   * resource is invalid, this will return PP_AUDIOSAMPLERATE_NONE.
   */
  PP_AudioSampleRate GetSampleRate(
      [in] PP_Resource config);

  /* Returns the sample frame count for the given AudioConfig resource. If the
   * resource is invalid, this will return 0. See RecommendSampleFrameCount for
   * more on sample frame counts.
   */
  uint32_t GetSampleFrameCount(
      [in] PP_Resource config);
};
