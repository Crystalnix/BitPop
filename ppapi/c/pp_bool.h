/* Copyright (c) 2010 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */
#ifndef PPAPI_C_PP_BOOL_H_
#define PPAPI_C_PP_BOOL_H_

#include "ppapi/c/pp_macros.h"

/**
 * @file
 * This file defines the PP_Bool enumeration for use in PPAPI C headers.
 */

/**
 *
 * @addtogroup Enums
 * @{
 */

/**
 * The PP_Bool enum is a boolean value for use in PPAPI C headers.  The
 * standard bool type is not available to pre-C99 compilers, and is not
 * guaranteed to be compatible between C and C++, whereas the PPAPI C
 * headers can be included from C or C++ code.
 */
typedef enum {
  PP_FALSE = 0,
  PP_TRUE = 1
} PP_Bool;
PP_COMPILE_ASSERT_SIZE_IN_BYTES(PP_Bool, 4);
/**
 * @}
 */

#endif  /* PPAPI_C_PP_BOOL_H_ */

