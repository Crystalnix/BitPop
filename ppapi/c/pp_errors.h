/* Copyright (c) 2011 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */
#ifndef PPAPI_C_PP_ERRORS_H_
#define PPAPI_C_PP_ERRORS_H_

/**
 * @file
 * This file defines an enumeration of all PPAPI error codes.
 */

/**
 *
 * @addtogroup Enums
 * @{
 */

/** This enumeration contains enumerators of all PPAPI error codes.
 * Errors are negative valued.
 */
enum {
  /**
   * This value is returned by a function on successful synchronous completion
   * or is passed as a result to a PP_CompletionCallback_Func on successful
   * asynchronous completion.
   */
  PP_OK = 0,

  /**
   * This value is returned by a function that accepts a PP_CompletionCallback
   * and cannot complete synchronously.  This code indicates that the given
   * callback will be asynchronously notified of the final result once it is
   * available.
   */
  PP_OK_COMPLETIONPENDING = -1,
  PP_ERROR_WOULDBLOCK = -1,  /* DEPRECATED: DO NOT USE */

  /** This value indicates failure for unspecified reasons. */
  PP_ERROR_FAILED = -2,

  /**
   * This value indicates failure due to an asynchronous operation being
   * interrupted, typically as a result of user action.
   */
  PP_ERROR_ABORTED = -3,

  /** This value indicates failure due to an invalid argument. */
  PP_ERROR_BADARGUMENT = -4,

  /** This value indicates failure due to an invalid PP_Resource. */
  PP_ERROR_BADRESOURCE = -5,

  /** This value indicates failure due to an unavailable PPAPI interface. */
  PP_ERROR_NOINTERFACE = -6,

  /** This value indicates failure due to insufficient privileges. */
  PP_ERROR_NOACCESS = -7,

  /** This value indicates failure due to insufficient memory. */
  PP_ERROR_NOMEMORY = -8,

  /** This value indicates failure due to insufficient storage space. */
  PP_ERROR_NOSPACE = -9,

  /** This value indicates failure due to insufficient storage quota. */
  PP_ERROR_NOQUOTA = -10,

  /**
   * This value indicates failure due to an action already being in
   * progress.
   */
  PP_ERROR_INPROGRESS = -11,

  /** This value indicates failure due to a file that does not exist. */
  PP_ERROR_FILENOTFOUND = -20,

  /** This value indicates failure due to a file that already exists. */
  PP_ERROR_FILEEXISTS = -21,

  /** This value indicates failure due to a file that is too big. */
  PP_ERROR_FILETOOBIG = -22,

  /**
   * This value indicates failure due to a file having been modified
   * unexpectedly.
   */
  PP_ERROR_FILECHANGED = -23,

  /** This value indicates failure due to a time limit being exceeded. */
  PP_ERROR_TIMEDOUT = -30,

  /** This value indicates that the user cancelled rather than providing
   * expected input.
   */
  PP_ERROR_USERCANCEL = -40
};

/**
 * @}
 * End of addtogroup Enums
 */

#endif  /* PPAPI_C_PP_ERRORS_H_ */
