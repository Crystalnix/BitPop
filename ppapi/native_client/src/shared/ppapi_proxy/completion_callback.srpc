# Copyright (c) 2010 The Native Client Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

# RPC method that supports invocation of plugin-side callbacks from the browser.
{
 'name': 'CompletionCallbackRpc',
 'rpcs': [
          {'name': 'RunCompletionCallback',
           'inputs': [['callback_id', 'int32_t'],
                      ['result', 'int32_t'],
                      ['read_buffer', 'char[]' ],
                     ],
           'outputs': []
          },
         ]
}
