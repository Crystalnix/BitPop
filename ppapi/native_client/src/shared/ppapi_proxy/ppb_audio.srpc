# Copyright (c) 2010 The Native Client Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

# This file declares the RPC methods used to implement PPB_Audio
# calls from the plugin.  The functions are described in
# ppapi/c/ppb_audio.h.
{
 'name': 'PpbAudioRpc',
 'rpcs': [
          # Implements a call to create an audio resource.
          # Returns descriptors for memory window, sync socket.
          {'name': 'PPB_Audio_Create',
           'inputs': [['instance', 'PP_Instance'],
                      ['config', 'PP_Resource'],
                     ],
           'outputs': [['out_resource', 'PP_Resource'],
                      ]
          },
          # Implements a call to determine IsAudio.
          {'name': 'PPB_Audio_IsAudio',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['out_bool', 'int32_t'],
                      ]
          },
          # Implements a call to get current config as resource.
          {'name': 'PPB_Audio_GetCurrentConfig',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['out_resource', 'PP_Resource'],
                      ]
          },
          # Implements a call to stop playback.
          {'name': 'PPB_Audio_StopPlayback',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['out_bool', 'int32_t'],
                      ]
          },
          # Implements a call to start playback.
          {'name': 'PPB_Audio_StartPlayback',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['out_bool', 'int32_t'],
                      ]
          },
         ]
}
