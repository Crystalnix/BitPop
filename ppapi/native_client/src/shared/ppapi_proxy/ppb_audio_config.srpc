# Copyright (c) 2010 The Native Client Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

# This file declares the RPC methods used to implement PPB_AudioConfig
# calls from the plugin.  The functions are described in
# ppapi/c/ppb_audio_config.h.
{
 'name': 'PpbAudioConfigRpc',
 'rpcs': [
          # Implements a call to create an audio configuration for stereo output.
          {'name': 'PPB_AudioConfig_CreateStereo16Bit',
           'inputs': [['instance', 'PP_Instance'],
                      ['sample_rate', 'int32_t'],
                      ['sample_frame_count', 'int32_t'],
                     ],
           'outputs': [['resource', 'PP_Resource'],
                      ]
          },
          # Implements a call to check if this is an AudioConfig.
          {'name': 'PPB_AudioConfig_IsAudioConfig',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['out_bool', 'int32_t']
                      ]
          },
          # Implements a call to get recommended sample frame count.
          {'name': 'PPB_AudioConfig_RecommendSampleFrameCount',
           'inputs': [['request_sample_rate', 'int32_t'],
                      ['request_sample_frame_count', 'int32_t'],
                     ],
           'outputs': [['out_sample_frame_count', 'int32_t']
                      ]
          },
          # Implements a call to get the sample rate.
          {'name': 'PPB_AudioConfig_GetSampleRate',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['sample_rate', 'int32_t']
                      ]
          },
          # Implements a call to get the obtained sample frame count.
          {'name': 'PPB_AudioConfig_GetSampleFrameCount',
           'inputs': [['resource', 'PP_Resource'],
                     ],
           'outputs': [['sample_frame_count', 'int32_t'],
                      ]
          },
         ]
}
