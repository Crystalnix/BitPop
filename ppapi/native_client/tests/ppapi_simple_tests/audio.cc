// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
//

#include <stdint.h>

#include <cmath>
#include <limits>
#include <string>

#include "native_client/src/shared/platform/nacl_check.h"
#include "native_client/src/shared/platform/nacl_log.h"


#include "ppapi/c/pp_bool.h"
#include "ppapi/c/pp_errors.h"
#include "ppapi/c/ppb_audio.h"
#include "ppapi/c/ppb_audio_config.h"
#include "ppapi/cpp/audio.h"
#include "ppapi/cpp/audio_config.h"
#include "ppapi/cpp/completion_callback.h"
#include "ppapi/cpp/instance.h"
#include "ppapi/cpp/module.h"

const PP_AudioSampleRate kSampleFrequency = PP_AUDIOSAMPLERATE_44100;
// Buffer size in units of sample frames.
// 4096 is a conservative size that should avoid underruns on most systems.
const uint32_t kSampleFrameCount = 4096;
const uint32_t kDefaultFrequency = 400;
const uint32_t kDefaultDuration = 10000;

const int kNumChannelsForStereo = 2;
const size_t kSizeSingleSample = kNumChannelsForStereo * sizeof(int16_t);

const double kPi = 3.141592653589;
const double kTwoPi = 2.0 * kPi;

class MyInstance : public pp::Instance {
 private:
  pp::Audio audio_;
  uint32_t obtained_sample_frame_count_;
  double audio_wave_time_;   // between -kTwoPi, +kTwoPi
  uint32_t duration_;   // in msec
  uint32_t frequency_;

  static void SoundCallback(void* samples, uint32_t num_bytes, void* thiz) {
    MyInstance* instance = reinterpret_cast<MyInstance*>(thiz);
    // CHECK inside callback is only for testing purposes.
    CHECK(instance->obtained_sample_frame_count_ * kSizeSingleSample ==
          num_bytes);
    const double delta = kTwoPi * instance->frequency_ / kSampleFrequency;
    const int16_t max_int16 = std::numeric_limits<int16_t>::max();
    int16_t* buf = reinterpret_cast<int16_t*>(samples);
    for (size_t i = 0; i < instance->obtained_sample_frame_count_; ++i) {
      const double v = sin(instance->audio_wave_time_) * max_int16;
      *buf++ = static_cast<int16_t>(v);
      *buf++ = static_cast<int16_t>(v);
      // Add delta, keep within -kTwoPi, +TwoPi to preserve precision.
      instance->audio_wave_time_ += delta;
      if (instance->audio_wave_time_ > kTwoPi)
        instance->audio_wave_time_ -= kTwoPi * 2.0;
    }
  }

  static void StopOutput(void* thiz, int32_t err) {
    if (PP_OK == err) {
      MyInstance* instance = static_cast<MyInstance*>(thiz);
      instance->audio_.StopPlayback();
    }
  }

  void ExtraChecks(pp::AudioConfig* config) {
    CHECK(obtained_sample_frame_count_ >= PP_AUDIOMINSAMPLEFRAMECOUNT);
    CHECK(obtained_sample_frame_count_ <= PP_AUDIOMAXSAMPLEFRAMECOUNT);

    PPB_GetInterface get_browser_if =
        pp::Module::Get()->get_browser_interface();

    const PPB_AudioConfig* audio_config_if =
        static_cast<const PPB_AudioConfig*>(
        get_browser_if(PPB_AUDIO_CONFIG_INTERFACE));

    const PPB_Audio* audio_if =
        static_cast<const PPB_Audio*>(
          get_browser_if(PPB_AUDIO_INTERFACE));

    CHECK(NULL != audio_config_if);
    CHECK(NULL != audio_if);

    const PP_Resource audio_config_res = config->pp_resource();
    const PP_Resource audio_res = audio_.pp_resource();

    CHECK(PP_TRUE == audio_config_if->IsAudioConfig(audio_config_res));
    CHECK(PP_TRUE == audio_if->IsAudio(audio_res));
    CHECK(PP_FALSE == audio_config_if->IsAudioConfig(audio_res));
    CHECK(PP_FALSE == audio_if->IsAudio(audio_config_res));
    CHECK(audio_if->GetCurrentConfig(audio_res) == audio_config_res);
    CHECK(0 == audio_if->GetCurrentConfig(audio_config_res));
    CHECK(audio_config_if->GetSampleRate(audio_config_res) ==
        config->sample_rate());
    CHECK(audio_config_if->GetSampleFrameCount(audio_config_res) ==
        config->sample_frame_count());
    CHECK(audio_.config().pp_resource() == audio_config_res);
  }

  void ParseArgs(uint32_t argc, const char* argn[], const char* argv[]) {
     for (uint32_t i = 0; i < argc; ++i) {
       const std::string tag = argn[i];
       if (tag == "duration") duration_ = strtol(argv[i], 0, 0);
       if (tag == "frequency") frequency_ = strtol(argv[i], 0, 0);
     }
  }

 public:
  explicit MyInstance(PP_Instance instance)
    : pp::Instance(instance),
      duration_(kDefaultDuration), frequency_(kDefaultFrequency) {}

  virtual ~MyInstance() {}

  virtual bool Init(uint32_t argc, const char* argn[], const char* argv[]) {
    ParseArgs(argc, argn, argv);
    obtained_sample_frame_count_ = pp::AudioConfig::RecommendSampleFrameCount(
      kSampleFrequency, kSampleFrameCount);

    pp::AudioConfig config =
      pp::AudioConfig(this, kSampleFrequency, obtained_sample_frame_count_);

    audio_ = pp::Audio(this, config, SoundCallback, this);

    ExtraChecks(&config);

    CHECK(audio_.StartPlayback());

    pp::CompletionCallback cc(StopOutput, this);
    pp::Module::Get()->core()->CallOnMainThread(duration_, cc, PP_OK);
    return true;
  }
};

// standard boilerplate code below
class MyModule : public pp::Module {
 public:
  virtual pp::Instance* CreateInstance(PP_Instance instance) {
    return new MyInstance(instance);
  }
};

namespace pp {
Module* CreateModule() {
  return new MyModule();
}

}  // namespace pp
