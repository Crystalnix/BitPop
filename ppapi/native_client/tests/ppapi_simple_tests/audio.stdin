# NOTE: to be used with a prolog from
#       tests/ppapi_proxy/sel_universal_ppapi_replay_prolog.stdin

# define a bunch of variables for readability
set_variable MODULE i(4444)
set_variable INSTANCE i(5555)
# to be safe we allocate a shmem region of 64k
set_variable SHM_SIZE 65536
set_variable NUM_FRAMES 4096
# size of the audio buffer = NUM_FRAMES * 2 (channels) * 2 (16bit audio)
set_variable BUF_SIZE 16384
#  SHM_UNUSED = SHM_SIZE - BUF_SIZE
set_variable SHM_UNUSED 49152

# arbitrary numbers we picked as handles
set_variable HANDLE_CFG i(30)
set_variable HANDLE_AUDIO i(34)

echo
echo "*** INIT MODULE"
# PPB_xxx replays below will be triggered as a reaction to PPP_InitializeModule
replay 1 PPB_GetInterface:s:i  s("PPB_Core;1.0") * i(1)
# The results contain a process id which is non-determinisitic.
# We must hide the results to support golden file diffing.
rpc PPP_InitializeModule hide-results i(0) ${MODULE} h(pepper_desc) s("${service_string}") * i(0) i(0)

echo
echo "*** INIT INSTANCE"
set_variable TAGS C(10,frequency\000)
set_variable VALUES C(4,800\000)
# The PPB_xxx replays below will be triggered as a reaction to
# PPP_Instance_DidCreate in roughly the order shown
replay 1 PPB_GetInterface:s:i s("PPB_AudioConfig;1.0") * i(1)
replay 1 PPB_AudioConfig_RecommendSampleFrameCount:ii:i i(44100) i(${NUM_FRAMES}) * i(${NUM_FRAMES})
replay 1 PPB_AudioConfig_CreateStereo16Bit:iii:i ${INSTANCE} i(44100) i(${NUM_FRAMES}) * ${HANDLE_CFG}
# TODO(robertm): explain why these are called three times
replay 3 PPB_Core_AddRefResource:i: ${HANDLE_CFG} *
replay 3 PPB_Core_ReleaseResource:i: ${HANDLE_CFG} *
replay 1 PPB_GetInterface:s:i s("PPB_Audio;1.0") * i(1)
replay 1 PPB_Audio_Create:ii:i ${INSTANCE} ${HANDLE_CFG} * ${HANDLE_AUDIO}
replay 1 PPB_AudioConfig_IsAudioConfig:i:i ${HANDLE_CFG} * i(1)
replay 1 PPB_Audio_IsAudio:i:i ${HANDLE_AUDIO} * i(1)
replay 1 PPB_AudioConfig_IsAudioConfig:i:i ${HANDLE_AUDIO} * i(0)
replay 1 PPB_Audio_IsAudio:i:i ${HANDLE_CFG} * i(0)
replay 1 PPB_Audio_GetCurrentConfig:i:i ${HANDLE_AUDIO} * ${HANDLE_CFG}
replay 1 PPB_Audio_GetCurrentConfig:i:i ${HANDLE_CFG} * i(0)
replay 1 PPB_AudioConfig_GetSampleRate:i:i  ${HANDLE_CFG} * i(44100)
replay 1 PPB_AudioConfig_GetSampleFrameCount:i:i  ${HANDLE_CFG} * i(${NUM_FRAMES})
replay 1 PPB_Audio_StartPlayback:i:i ${HANDLE_AUDIO} * i(1)
replay 1 PPB_Core_CallOnMainThread:iii: i(10000) i(1) i(0) *
rpc PPP_Instance_DidCreate ${INSTANCE} i(1) ${TAGS} ${VALUES} * i(0)

echo
echo "*** TRIGGER REPAINT"
rpc PPP_Instance_DidChangeView ${INSTANCE} I(4,8,79,400,400) I(4,0,0,400,400) *

echo
echo "*** SETTING UP AUDIO SHARED MEMORY"
shmem audio_handle audio_address ${SHM_SIZE}
memset ${audio_address} 0  ${SHM_SIZE} 0
checksum ${audio_address} 0 ${SHM_SIZE}

echo
echo "*** SETTING UP AUDIO SYNC SOCKET"
sync_socket_create sync_in sync_out

echo
echo "*** AUDIO STREAM"
# inform the nexe about the shmem region and the sync_socket
rpc PPP_Audio_StreamCreated ${HANDLE_AUDIO} h(audio_handle) i(${BUF_SIZE}) h(sync_out) *

echo
echo "*** AUDIO STREAM SYNC"
# By writing a byte into the sync sock we wake up the other side, the actual
# value does not matter as long as it is not -1.
# once the other side has woken up it is supposed to write into the buffer
# associated with audio_handle as quickly as possible.
# We wait 1sec to give it time to do that.
# Note: the shmem area is actually bigger than the buffer.
# We compute a checksum over the part that is used and that isn't.
# The latter checksum should not change and be zero.
sync_socket_write sync_in 0
sleep 1
checksum ${audio_address} 0 ${BUF_SIZE}
checksum ${audio_address} ${BUF_SIZE} ${SHM_UNUSED}

echo
echo "*** AUDIO STREAM SYNC"
sync_socket_write sync_in 1
sleep 1
checksum ${audio_address} 0 ${BUF_SIZE}
checksum ${audio_address} ${BUF_SIZE} ${SHM_UNUSED}

echo
echo "*** AUDIO STREAM SYNC"
sync_socket_write sync_in 2
sleep 1
checksum ${audio_address} 0 ${BUF_SIZE}
checksum ${audio_address} ${BUF_SIZE} ${SHM_UNUSED}

echo
echo "*** AUDIO STREAM SYNC"

sync_socket_write sync_in 3
sleep 1
checksum ${audio_address} 0 ${BUF_SIZE}
checksum ${audio_address} ${BUF_SIZE} ${SHM_UNUSED}

echo
echo "*** TERMINATE AUDIO STREAM"
sync_socket_write sync_in -1
checksum ${audio_address} 0 ${BUF_SIZE}
checksum ${audio_address} ${BUF_SIZE} 49152

sleep 1
rpc PPP_ShutdownModule *
