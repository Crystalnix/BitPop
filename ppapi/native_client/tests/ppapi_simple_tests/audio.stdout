
"*** REPLAY PROLOG"

"*** REGISTER REPLAY METHODS"

"*** CREATE UPCALL SERVICE"

"*** INIT MODULE"
rpc call intiated PPP_InitializeModule:iihs:ii
input 0:  i(0)
input 1:  i(4444)
input 2:  h(pepper_desc)
input 3:  @LENGTHY_SERVICE_STRING
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Core;1.0")
output 0:  i(1)
rpc call complete PPP_InitializeModule:iihs:ii


"*** INIT INSTANCE"
rpc call intiated PPP_Instance_DidCreate:iiCC:i
input 0:  i(5555)
input 1:  i(1)
input 2:  C(10,frequency\x00)
input 3:  C(4,800\x00)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_AudioConfig;1.0")
output 0:  i(1)
replaying PPB_AudioConfig_RecommendSampleFrameCount:ii:i:
input 0:  i(44100)
input 1:  i(4096)
output 0:  i(4096)
replaying PPB_AudioConfig_CreateStereo16Bit:iii:i:
input 0:  i(5555)
input 1:  i(44100)
input 2:  i(4096)
output 0:  i(30)
replaying PPB_Core_AddRefResource:i::
input 0:  i(30)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Audio;1.0")
output 0:  i(1)
replaying PPB_Audio_Create:ii:i:
input 0:  i(5555)
input 1:  i(30)
output 0:  i(34)
replaying PPB_Core_AddRefResource:i::
input 0:  i(30)
replaying PPB_Core_ReleaseResource:i::
input 0:  i(30)
replaying PPB_AudioConfig_IsAudioConfig:i:i:
input 0:  i(30)
output 0:  i(1)
replaying PPB_Audio_IsAudio:i:i:
input 0:  i(34)
output 0:  i(1)
replaying PPB_AudioConfig_IsAudioConfig:i:i:
input 0:  i(34)
output 0:  i(0)
replaying PPB_Audio_IsAudio:i:i:
input 0:  i(30)
output 0:  i(0)
replaying PPB_Audio_GetCurrentConfig:i:i:
input 0:  i(34)
output 0:  i(30)
replaying PPB_Audio_GetCurrentConfig:i:i:
input 0:  i(30)
output 0:  i(0)
replaying PPB_AudioConfig_GetSampleRate:i:i:
input 0:  i(30)
output 0:  i(44100)
replaying PPB_AudioConfig_GetSampleFrameCount:i:i:
input 0:  i(30)
output 0:  i(4096)
replaying PPB_Audio_StartPlayback:i:i:
input 0:  i(34)
output 0:  i(1)
replaying PPB_Core_CallOnMainThread:iii::
input 0:  i(10000)
input 1:  i(1)
input 2:  i(0)
replaying PPB_Core_ReleaseResource:i::
input 0:  i(30)
rpc call complete PPP_Instance_DidCreate:iiCC:i
output 0:  i(1)


"*** TRIGGER REPAINT"
rpc call intiated PPP_Instance_DidChangeView:iII:
input 0:  i(5555)
input 1:  I(4,8,79,400,400)
input 2:  I(4,0,0,400,400)
rpc call complete PPP_Instance_DidChangeView:iII:


"*** SETTING UP AUDIO SHARED MEMORY"
CHECKSUM: 0x00000000

"*** SETTING UP AUDIO SYNC SOCKET"

"*** AUDIO STREAM"
rpc call intiated PPP_Audio_StreamCreated:ihih:
input 0:  i(34)
input 1:  h(audio_handle)
input 2:  i(16384)
input 3:  h(sync_out)
rpc call complete PPP_Audio_StreamCreated:ihih:


"*** AUDIO STREAM SYNC"
CHECKSUM: 0xbbf21e00
CHECKSUM: 0x00000000

"*** AUDIO STREAM SYNC"
CHECKSUM: 0xcbef91ac
CHECKSUM: 0x00000000

"*** AUDIO STREAM SYNC"
CHECKSUM: 0xd0f22174
CHECKSUM: 0x00000000

"*** AUDIO STREAM SYNC"
CHECKSUM: 0x23742c68
CHECKSUM: 0x00000000

"*** TERMINATE AUDIO STREAM"
CHECKSUM: 0x23742c68
CHECKSUM: 0x00000000
rpc call intiated PPP_ShutdownModule::
rpc call complete PPP_ShutdownModule::

